### ISSUE TITLE
Reworking the implementation of neat_set_property

### ISSUE THREAD

**Issue by oystedal** [NEAT-project/neat#81]
As of currently, the properties requested by an application is specified as a bit-mask inside a single integer, which is then passed to `neat_set_property()`. I propose to use strings to specify requests for features instead of using a bit-mask.

Example usage would be `neat_set_property(ctx, flow, "TCP_REQUIRED")` and  `neat_set_property(ctx, flow, "IPV6_REQUIRED")`.

This is less efficient than using a bit-mask or a larger bitmap, but I still believe this can be made reasonably efficient using e.g. a hashmap. A string carries more meaningful information on its own, which is useful during debugging or logging.

I'm interested in hearing opinions on this matter.

**Comment by bozakov** [comment]
I'm very much in favor of this change, and have two comments:
1. In Aberdeen we agreed that we need a way to specify priorities to properties (i.e., optional, mandatory). 
2. I think we need key/value pairs, e.g., `tcp_required:true` in order to support more advanced properties further on (e.g. available bandwidth=50 MBps) 

As I recall we also agreed to use JSON as the format to exchange the properties with the policy manager. 

I'll upload our initial policy manager prototype with the documentation of the properties as we envision them later today.

**Comment by oystedal** [comment]
Hi, and thank you for the comments.

1. I'll get someone to bring me up to date on this point as I was not present in Aberdeen.
2. We also agreed here that a key/value approach would be useful and possibly even required, so this is already a work in progress.

As a side note: having the policy manager, even if it's just a small prototype, is extremely useful to ensure that the code we implement here will be compatible with the policy manager, so this is very appreciated.

**Comment by kristrev** [comment]
I am not so much in favor of pure strings, as it will be difficult to manage, keep track of an use as the number of properties increases. For example, the strings would need to have a fixed structure (or at least should how) to make them usable and the documentation understandable for other people than those that write the code.

I propose that we we pass to set_property() is a JSON object (as a string of course) that can contain one or more properties. This will take care of the structure requirement, key/value and it will ideally reduce the complexity of the policy manager, as we can hopefully just take the JSON output from the policy manager and pass it to the set_property() call. If we go with pure strings, we would need to translate the policy-objects into strings strings. This strikes me as error-prone, boring, redundant and all-in-all, monkey work.

**Comment by kristahi** [comment]
@kristrev Well, we will need to keep track of valid properties somehow in any case. And indeed, a mapping to string names for serialization to JSON that can be exchanged with the PM.

I am not very much in favour of setting properties as JSON in the C API, that will get even more messy than just using strings IMHO. Maybe I misunderstood what you meant?

**Comment by kristrev** [comment]
Yes, documentation is needed, but my experience is that JSON is much easier to work with for most people than pure strings. For example, that we get the structure for free (instead of an artificial X_Y_Z rule for strings) is a huge win and will make the API easier to work with.

No properties will be set as JSON, but a valid JSON-object will be passed to and parsed by the NEAT library. This is fairly efficient and trivial code, and the input object can be generated by code written in any language (or defined as a string constant by whoever uses the NEAT library). Constants will be used anyway (no-one will ever write the strings directly) so I don't really see the problem. Using JSON also enables to easily set multiple properties at once.

However, if JSON is off the stable, I am still very much against using strings. It would instead make more sense to base set_/get_property() on set/getsockopt() and a level/optname approach (where optname is an int).

**Comment by tuexen** [comment]
I agree with Kristian (kristrev ).

**Comment by oystedal** [comment]
The approach we're currently working on exposes several functions for different type of values, that is, neat_property_[get|set]_[bool|int|float|string], which takes a string as parameter for the property name and and a parameter for the actual value with the appropriate C type, e.g. `neat_set_property_bool(flow, "tcp_required", 0);`

I agree with @kristahi that an application should not have to deal with JSON strings/objects in order to configure the framework. A get/setsockopt-inspired approach is a far better alternative than JSON in my opinion. 

Since we have to work with JSON internally, we can always expose part of this functionality to allow applications to set multiple options when this is desired.

**Comment by mangis-emc** [comment]
Why are you working on an "approach" to policy and properties? This integration topic should be discussed in a plenary meeting, or at least with most of the code/policy people involved.

IMHO a JSON file is the best way to configure a framework. A user can easily specify parameters and settings. I mean the final end user. Then this could be always translated to a bitmask or flags in the core layer.

We also need to agree on a common "property dictionary" that the NEAT logic core has to understand to setup sockets and connections.

**Comment by bozakov** [comment]
@oystedal I have uploaded the first version of the policy manager code to the repo. It's a bit rough but should be enough to get an idea of the policy and property logic. I'll be extending the walkthrough  documentation in the coming days. Please let me know if you have any questions about the code.

@kristrev I am in favour of using JSON as the exchange format.

**Comment by mangis-emc** [comment]
Test.

Simone Mangiante | Senior Research Scientist | EMC Information Systems International
Ph. +353 21 4281761 (direct) +353 (0) 834582892 (mobile) | simone.mangiante@emc.com<mailto:simone.mangiante@emc.com>
Business Address: Ireland COE | Ovens | Co Cork | Ireland

[cid:image001.png@01CF79B1.0CCA23A0]
EMC Information Systems International
Registered in Ireland with Company Number: 461022 and Registered address: IDA Industrial Estate, Ovens, Co. Cork
The information contained in this e-mail message and any files transmitted with it are confidential. It is intended only for the addressee and others authorised to receive it. If you are not the intended recipient or the person responsible for delivering the message to the intended recipient, you are advised that you have received the e-mail in error; please delete it and notify the sender immediately. You should not retain the message or disclose its contents to anyone. Any disclosure, copying, distribution or action taken in reliance on the contents of the e-mail and its attachments is strictly prohibited.

**Comment by naeemk** [comment]
This is a conversation we anyhow needed to have and I'm glad it's happening now as we begin to develop the NEAT logic code. @kristrev 's suggestion seem reasonable to me (to use JSON) and better compatible with PM's current code.

**Comment by naeemk** [comment]
@mangis-emc : if no objection to above comment and no necessity to discuss this in plenary, I can "close" the issue.

**Comment by kristrev** [comment]
Since the policy manager and the properties will be tightly coupled, I assumed you had discussed closely with EMC before starting anything (and would include Celerway at some point). I was a bit surprised that was not the case, and I agree with @mangis-emc that the properties, the API, etc. needs to be discussed in detail with the other involved parties. The plenary is the correct place to do this and otherwise we risk redundant work, implementing something that will not work and has to be redone, conflicts, etc. In order to avoid this, I suggest we postpone this (get_/set_property()) until after the plenary in Oslo.

Instead, it would be very useful if UiO could look at the "Callbacks for D1.2 events" AP. This is much more important for NEAT, as it is the last missing piece of what I would call the NEAT foundation (well, working HE would also be nice). It also looks weird that the code and what has been written in the deliverable are completely different. If I was a reviewer, I would ask what the point is of D1.2 when the code doesn't follow it anyway.

Even though the PM and properties belong to the NEAT core, they belong, in my view, to NEAT v0.2. Getting the API in sync with D1.2 (+ basic HE) is a must for v0.1. The properties we have now are fine for an initial version.

**Comment by naeemk** [comment]
> Since the policy manager and the properties will be tightly coupled, I assumed you had discussed closely with EMC before starting anything (and would include Celerway at some point). I was a bit surprised that was not the case

Nothing has been started, on his first day of work, he's making a suggestion and asking for feedback as I read above and the git is the right place to do that :-).

> , and I agree with @mangis-emc that the properties, the API, etc. needs to be discussed in detail with the other involved parties. The plenary is the correct place to do this and otherwise we risk redundant work, implementing something that will not work and has to be redone, conflicts, etc. In order to avoid this, I suggest we postpone this (get_/set_property()) until after the plenary in Oslo.

I see your point. However, please clearly define what part of D1.2 apart than the list of properties you would like to discuss in the plenary and what parts we can work on since you want "the API" to be discussed in details in plenary while most of the NEAT Logic *is* in fact the API, except what is the initial "callback-based NEAT API framework"   

>Instead, it would be very useful if UiO could look at the "Callbacks for D1.2 events" AP. This is much more important for NEAT, as it is the last missing piece of what I would call the NEAT foundation (well, working HE would also be nice). It also looks weird that the code and what has been written in the deliverable are completely different. If I was a reviewer, I would ask what the point is of D1.2 when the code doesn't follow it anyway.

>Even though the PM and properties belong to the NEAT core, they belong, in my view, to NEAT v0.2. Getting the API in sync with D1.2 (+ basic HE) is a must for v0.1. The properties we have now are fine for an initial version.

We're working on that too. 

Naeem

**Comment by kristrev** [comment]
Properties and policies are tightly coupled and needs to be properly discussed and an overall design needs to be ready before we start writing code. This is something that needs to be done face to face, no amount of emails, chats or telcos can replace a blackboard when it comes to design. Worst-case, if properties are not done right, the policy manager will have to circumvent them and we will end up with two large code components that will do roughly the same thing.

I don't want to discuss anything related to D1.2 in the plenary, implementing the D1.2 callbacks should be ready before the plenary as per the action points from Aberdeen. And as far as i can see, get/set_property() is only briefly mentioned on page 13 of D1.2, while there are several function calls to that have a very clear definition. So it would make much more sense to start with those (they seem to be enough work), while we work out properties and policies since that is a lot more complicated.

**Comment by naeemk** [comment]
> With the API, I meant the function calls related to the get/set_property(). Properties and policies are tightly coupled and needs to be properly discussed and an overall design needs to be ready before we start writing code. This is something that needs to be done face to face, no amount of emails, chats or telcos can replace a blackboard when it comes to design. Worst-case, if properties are not done right, the policy manager will have to circumvent them and we will end up with two large code components that will do roughly the same thing.

Sure, I have made a raw entry to https://secure.mlab.no/neatwiki/doku.php?id=oslo-june-2016-agenda.

**Comment by tuexen** [comment]
Regarding fixing the HE code: I assume Karl-Johan  is going to fix it. Last time Felix/I pinged him,
he was on a paper deadline which was last Sunday. However, that deadline was extended by a week. So hopefully he can fix it next week. I would also like to add the UDP encaps stuff, but I need green buildbots for that to make sure the stuff runs on all platforms and it will increase the number of candidates. We will also then add some more test cases to cover more HE scenarios, especially ones with non-working candidates...

**Comment by mangis-emc** [comment]
@naeemk thank you for the item you put in the agenda for Oslo. We could have a discussion this month before the plenary meeting, in order to be prepared and more focused there. Is the wiki the right place?

**Comment by naeemk** [comment]
@mangis-emc: sounds a very good idea to me! we can sketch some ideas for the "list of properties" and the interaction of API with PM (e.g. JSON syntax), in a wiki page and discuss them among us (UiO+EMC and whoever else that might be interested) over a telco when the page is populated, all before the plenary => so to use the plenary time in a more efficient way.
